rcmgr.go

```diff
@@ -0,0 +1,441 @@
+package rcmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+type ResourceManager struct {
+	limits Limiter
+
+	system    *SystemScope
+	transient *TransientScope
+
+	mx    sync.Mutex
+	svc   map[string]*ServiceScope
+	proto map[protocol.ID]*ProtocolScope
+	peer  map[peer.ID]*PeerScope
+
+	cancelCtx context.Context
+	cancel    func()
+	wg        sync.WaitGroup
```

Comment:
Nit: move these 3 member above the mutex (to clarify that the mutex only protects the maps).

Reply:
sure, mutex hat ftw.

Reply:
done